### ISSUE TITLE
Success Criterion 1.4.11 - Non-text Contrast - Level AA

### ISSUE THREAD

**Issue by JJdeGroot** [w3c/matf#49]
WCAG2ICT guidance: https://www.w3.org/TR/wcag2ict-22/#non-text-contrast

Share your thoughts for applying to mobile apps as a comment below.

**Comment by qbalsdon** [comment]
I think we need to align with the regular guidance, if we take external factors like sunshine into consideration we could be far more prohibitive and not actually very helpful to the end user

**Comment by JJdeGroot** [comment]
<details>
<summary>MATF meeting on July 31, 2024</summary>

[Source](https://www.w3.org/2024/07/31-matf-minutes.html#t03)
JJ: Reading text from WCAG2ICT, mentioning the user-agent exception for contrast needs to be mapped to mobile.

<Mick> +1 to the inactive component exception being problematic.

<Jamie> did we skip 1.4.10 REflow?

<Joe_Humbert> +1 to the inactive component exception being problematic

<JJ> 1.4.10 discussion @Jamie: [w3c/matf#4](https://github.com/w3c/matf/issues/4)

Joe_Humbert: About not modified by the author. This doesn't seem to be defined - what does that mean? For example, if the background colour is changed but not the component, does that not impact the contrast?

<julianmka> Agree with that - thinking about iOS default palette and text styles that don't have sufficient contrast

<quintinb> +1 we should not rely on manufacturers being accessible. We need to hold them accountable too

@JJ we should create a separate ticket about user agent

[ACTION: Github issue for User-Agent discussion](https://github.com/w3c/matf/issues/63)
</details>

**Comment by Keanem6** [comment]
Should we be include requirements for Dark mode as well?

**Comment by JJdeGroot** [comment]
Example of component which is "not modified by the author"

Light gray: #EAEAEB on white #FFF = 1,20:1 contrast
![IMG_9646](https://github.com/user-attachments/assets/dc4616a2-8555-4b68-9a94-d92c553d8c11)

After enabling high contrast

Darker gray: #96969d on white #FFF = 2,94:1 contrast
![IMG_9647](https://github.com/user-attachments/assets/05855f11-da88-4cc2-8131-d89ce753e775)

This might actually have 3:1 contrast because the screenshot compression reduces the color accuracy by 1-10%
(that's one of the problems with calculating contrast of mobile apps when you don't have source code access)

**Comment by JJdeGroot** [comment]
A few notes/comments below from e-mail conversation with @alastc (AC)

JJ:
For 1.4.11 Non-text Contrast, there is an exception for components where the appearance is determined by the user agent AND not modified by the author
- Where do we draw the line for components provided on Android and iOS? 
- For example, Jetpack Compose is a declarative UI framework which requires developers to add extra dependencies. 
- To add Material Design components, you also need to add extra dependencies. Do we still consider these to be provided by the user-agent when they are developed by Google?
- It can be difficult to check whether a component was modified by the author, because auditors usually don't have access to the source code. E.g. it's not easy to "inspect element" to check for CSS like you can do on the web.

AC: I don’t think any of these would count as provided by the “user agent”, as there isn’t a user-agent involved for native apps (unless something were provided by the AT, such as the bounding box for focused items).

**Comment by JJdeGroot** [comment]
From today's meeting, @detlevhfischer mentioned: "ACTION: Add agenda item for contrast exception for OS-driven keyboard focus". We will consider this in the next meeting that will cover 1.4.11

**Comment by julianmka** [comment]
Default iOS components, in particular, tend to fail this SC. Thinking about the ubiquitous chevron on grouped rows -- the standard gray used there doesn't meet color contrast requirements. This is another SC where we may need to acknowledge known failures.

**Comment by JJdeGroot** [comment]
Discussed in today’s meeting.

<details>
<summary>18 December 2024</summary>

[Source](https://www.w3.org/2024/12/18-matf-minutes.html#c13a)

~~~text
Joe_Humbert: Summarizes previous discussion on 1.4.11.

Joe_Humbert: Should we address Dark Mode as a separate item or as part of each relevant SC?

Illai: Include it in system settings.

Mick: Agree that we should call it out in system settings guidance but also on relevant SCs.

Joe_Humbert: Highlights conversation about the meaning of "user agent" in the native context ([w3c/matf#49 (comment)](https://github.com/w3c/matf/issues/49#issuecomment-2324142511)). This could be a high-level definition used in notes for relevant SCs.

<Mick> [w3c/matf#63](https://github.com/w3c/matf/issues/63) - user agent definition issue

Joe_Humbert: The definition of "user interface component" changed in a recent update to WCAG 2.2; we should find out what changed.
… after looking, perhaps there wasn't a change. We can't find a difference between 2.1 and 2.2 definitions.

Mick: Re: contrast on keyboard focus, we don't test for that internally due to complexity and difficulty.
… I don't think Deque does, either.

Jamie: Users can change the default keyboard focus indicator color, thickness, and other features in iOS. Android users don't have that option unless they're also using TalkBack with the external keyboard.

Joe_Humbert: For Android, that's not changing the keyboard focus but screen reader focus style.

Joe_Humbert: Speaking to Android keyboard focus, I know of an issue that was filed where Google closed it as won't fix.

Illai: We are covering focus indicators which isn't related to non-text contrast. Other SCs cover that.

<Mick> Deque testing methodology also don't test for visible focus color changes - although they focus their testing around switch control and state focus indicator that cannot be overridden (for iOS and Android).

julianmka: Seems to be agreement that this SC applies to native mobile, what would we need to address in notes as caveats?

Jon_Gibbins: Non-text contrast applies to keyboard at AA because Focus Appearance is AAA. While other SCs also relate to keyboard accessibility, Focus Appearance is separate from contrast requirements. Focus Appearance is more about how easily discernable focus style is.
… In my view, this SC (1.4.11) requires a certain level of contrast for keyboard focus indicator

Jamie: Could we add a note about the user customizing the focus indicator, do we consider dark mode, and do we need to address hover/focus states?

<Jon_Gibbins> Thanks everyone

ACTION: Consider Dark Mode in the bigger note about respecting system settings.
~~~
</details>

- ACTION: Consider Dark Mode in the bigger note about respecting system settings.

**Comment by JJdeGroot** [comment]
Discussed in today’s meeting.

<details>
<summary>16 April 2025</summary>

[Source](https://www.w3.org/2025/04/16-matf-minutes.html#b5a3)

~~~text
1.4.11 Non-text Contrast

<JJ> [w3c/matf#49](https://github.com/w3c/matf/issues/49)

<pauljadam> iOS Toggles have insufficient contrast ratio but their appearance is determined by the user agent and not modified by the author. You can add a custom .toggleStyle to achieve 3:1 contrast ratio but that is only required when the author modifies the design of the toggle.

I wonder how non-disabled folks would react if not-enabled components where just black text on blackgrounds...

is Flutter and React native a user agent or an author?

<JJ> User agent definition: [w3c/matf#63](https://github.com/w3c/matf/issues/63)

<JJ> pauljadam

pauljadam the ios toggle has insufficient contrast ratio - appearence is determined by user agent, but can be configured. If cross platform do their own thing, then they're on the hook for sufficient contrast

I wonder what the point of a disabled component is if it doesn't have to meet contrast requirements

<JJ> For frameworks like Flutter or React Native, could look purely at the colors, and not at technical implementation for 1.4.11

Joe_Humbert user agent: WCAG2ICT: any software that retrieves and represents documents for users

Not pages?

or views?

<Joe_Humbert> yup

<pauljadam> page is a document :)

<Joe_Humbert> https://www.w3.org/TR/wcag2ict-22/#user-agent

Haha sorry just stirring

<Joe_Humbert> Document is defined as well https://www.w3.org/TR/wcag2ict-22/#document

pauljadam another weird thing about toggles: off state is not sufficient, but on state does - is this a way to weasel out of this? because one state is passable?

<Detlev> I created an issue for redundant placeholder contrast requirement: [w3c/wcag#4343](https://github.com/w3c/wcag/issues/4343)

If there are some components that "don't matter" - what's the point of requiring anything? Maybe that's me just being synical

<pauljadam> I'm still surprised there is no definition of "inactive"

Tanya I think if you look at the intent of the SC is to make elements more detectable for people. I would not mind adding a note that all elements should comply with minimum contrast

By allowing low contrast for inactive elements, we're telling those who can't see it "you wouldn't get it, your understanding / perception doesn't matter"

<rachaely> does the rounding note also apply for text contrast or just non-text?

<pauljadam> on iOS when you enable increase contrast setting the toggle then has better contrast too. so we may not want to put a burden on developers that they have to fix all of apple's bad contrast choices

JJ maybe add a note to modify the appearance for accessibility, that authors should do it

<pauljadam> Or apple could fix their default contrast then no developers would have to worry about it 🤷‍♂️

ACTION: Determine potential note to cover best-practices for user agent determined appearance and inactive components

pauljadam maybe we should ask developers to file bug reports

ACTION: Note could also say something about "if you have the ability to change the appearace to reach minimum contrast, then do it"

<quintinb> +1 pauljadam

"Email your MP"

I may sound like a lazy engineer, but I'm not going to lie, I think the OS should do it
~~~
</details>

- ACTION: Determine potential note to cover best-practices for user agent determined appearance and inactive components
- ACTION: Note could also say something about "if you have the ability to change the appearance to reach minimum contrast, then do it"

**Comment by JJdeGroot** [comment]
Related: https://github.com/w3c/wcag/issues/4290#issuecomment-2761758769

**Comment by yakimvanzuijlen** [comment]
One thing that WCAG does well is encouraging authors to use standardized components whenever possible. This way only a few OS vendors need to work on creating robust components instead of requiring every author to create their own robust and accessible UI framework. Even on the web, components often have had accessibility issues (things such as focus rings and input elements come to mind). But the value of having these components standardized is important in the long term. There's a lot of value in having authors rely on system-provided components. Users experience consistent, familiar, and predictable behavior. When system-provided components get updated all apps that use them benefit. With standards like EN 301 549 OS vendors are now also held to a higher standard to make sure their components meet accessibility requirements.

> "In a native app, the developer puts things together (including, optionally, out-of-the-box components) and sends that to the user. The user gets that software, unmodified by any intermediate software."

Especially with declarative UI frameworks this is less the case. The whole point of declarative languages is that the author specified the intent but leaves it up to the system how the components look and function. This is very similar to the declarative approach in HTML. There might be some APIs exposed that allow you to further tweak certain parameters, but this is not always the case (such as for the background color of a switch).

> "E.g. when a web developer puts a select-box on the page, the dev doesn’t control the behaviour (or much of the appearance) of the select-box. It is provided by the browser. I think that’s a different scenario from native."

I don't think it is that different. I only have experience with SwiftUI, but Jetpack Compose seems to have a similar approach. Take for example the way a toggle is authored in HTML and SwiftUI for example:
```html
<label>Receive newsletter<input type="checkbox" id="newsletterInput"></label>
```
```swift
Toggle("Receive newsletter", isOn: $newsletterInput)
```
In both instances there is only limited ability to style the system-provided checkbox component. On the web there is the `accent-color` CSS property and in SwiftUI there is the equivalant `tint()` view modifier. Granted, SwiftUI does allow you to create a fully custom design for a checkbox using the `ToggleStyle` protocol, ut that is not the case for all components (such as the standard date picker). Even then the underlying behavior and accessibility element representation is unchanged. In a native app you can build your own custom checkbox component, but how is that different from the web, where you could build a custom checkbox using javascript and ARIA?

As mentioned, there are still caveats with how to deal with contrast of unmodified components against an author-modified background color, but those issues also exist on the web.

In conclusion, I do think that it is still a good idea to apply "appearance of the component is determined by the [x] and not modified by the author" to the native mobile context. Although "user agent" should probably be changed.
